schemaregistry/rest_service.go

```diff
@@ -0,0 +1,323 @@
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// Relative Confluent Schema Registry REST API endpoints as described in the Confluent documentation
+// https://docs.confluent.io/current/schema-registry/docs/api.html
+const (
+	base              = ".."
+	schemas           = "/schemas/ids/%d"
+	schemasBySubject  = "/schemas/ids/%d?subject=%s"
+	subject           = "/subjects"
+	subjects          = subject + "/%s"
+	subjectsNormalize = subject + "/%s?normalize=%t"
+	version           = subjects + "/versions"
+	versionNormalize  = subjects + "/versions?normalize=%t"
+	versions          = version + "/%v"
+	compatibility     = "/compatibility" + versions
+	config            = "/config"
+	subjectConfig     = config + "/%s"
+	mode              = "/mode"
+	modeConfig        = mode + "/%s"
+)
+
+// REST API request
+type api struct {
+	method    string
+	endpoint  string
+	arguments []interface{}
+	body      interface{}
+}
+
+// newRequest returns new Confluent Schema Registry API request */
+func newRequest(method string, endpoint string, body interface{}, arguments ...interface{}) *api {
+	return &api{
+		method:    method,
+		endpoint:  endpoint,
+		arguments: arguments,
+		body:      body,
+	}
+}
+
+/*
+* HTTP error codes/ SR int:error_code:
+*	402: Invalid {resource}
+*	404: {resource} not found
+*		- 40401 - Subject not found
+*		- 40402 - SchemaMetadata not found
+*		- 40403 - Schema not found
+*	422: Invalid {resource}
+*		- 42201 - Invalid Schema
+*		- 42202 - Invalid SchemaMetadata
+*	500: Internal Server Error (something broke between SR and Kafka)
+*		- 50001 - Error in backend(kafka)
+*		- 50002 - Operation timed out
+*		- 50003 - Error forwarding request to SR leader
+ */
+
+// RestError represents a Schema Registry HTTP Error response
+type RestError struct {
+	Code    int    `json:"error_code"`
+	Message string `json:"message"`
+}
+
+// Error implements the errors.Error interface
+func (err *RestError) Error() string {
+	return fmt.Sprintf("schema registry request failed error code: %d: %s", err.Code, err.Message)
+}
+
+type restService struct {
+	url     *url.URL
+	headers http.Header
+	*http.Client
+}
+
+// newRestService returns a new REST client for the Confluent Schema Registry
+func newRestService(conf *kafka.ConfigMap) (*restService, error) {
+	urlConf, err := conf.Get("schema.registry.url", "")
+
+	if err != nil {
+		return nil, err
+	}
+
+	u, err := url.Parse(urlConf.(string))
+
+	if err != nil {
+		return nil, err
+	}
+
+	headers, err := newAuthHeader(u, conf)
+	if err != nil {
+		return nil, err
+	}
+
+	headers.Add("Content-Type", "application/vnd.schemaregistry.v1+json")
+	if err != nil {
+		return nil, err
+	}
+
+	transport, err := configureTransport(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	timeout, err := conf.Get("request.timeout.ms", 10000)
```

Comment:
is the intent to use the producer config of the same name? I think having a separate config value is more appropriate (with different name if they're intended to be in the same namespace).